In [1]:
import spacy
nlp = spacy.load('en')
from spacy.lang.en import English
#parser = English()
#nlp = spacy.load('en_core_web_sm')
example='''
That this House notes the announcement of 300 redundancies at the Nestlé manufacturing factories in York, Fawdon, Halifax and Girvan and that production of the Blue Riband bar will be transferred to Poland; acknowledges in the first three months of 2017 Nestlé achieved £21 billion in sales, a 0.4 per cent increase over the same period in 2016; further notes 156 of these job losses will be in York, a city that in the last six months has seen 2,000 job losses announced and has become the most inequitable city outside of the South East, and a further 110 jobs from Fawdon, Newcastle; recognises the losses come within a month of triggering Article 50, and as negotiations with the EU on the UK leaving the EU and the UK's future with the EU are commencing; further recognises the cost of importing products, including sugar, cocoa and production machinery, has risen due to the weakness of the pound and the uncertainty over the UK's future relationship with the single market and customs union; and calls on the Government to intervene and work with hon. Members, trades unions GMB and Unite and the company to avert these job losses now and prevent further job losses across Nestlé.'''

In [10]:
def entities(example, show=True):
    if show: print(example)
    #parsedEx = parser(example)
 
    print("-------------- entities only ---------------")
    # if you just want the entities and nothing else, you can do access the parsed examples "ents" property like this:
    doc = nlp(example)
    ents=doc.ents
    print(ents)
    tags={}
    for entity in ents:
        #print(entity.label, entity.label_, ' '.join(t.orth_ for t in entity))
        term=' '.join(t.orth_ for t in entity)
        if ' '.join(term) not in tags:
            tags[term]=[(entity.label, entity.label_)]
        else:
            tags[term].append((entity.label, entity.label_))
    print(tags)

In [11]:
entities(example)


That this House notes the announcement of 300 redundancies at the Nestlé manufacturing factories in York, Fawdon, Halifax and Girvan and that production of the Blue Riband bar will be transferred to Poland; acknowledges in the first three months of 2017 Nestlé achieved £21 billion in sales, a 0.4 per cent increase over the same period in 2016; further notes 156 of these job losses will be in York, a city that in the last six months has seen 2,000 job losses announced and has become the most inequitable city outside of the South East, and a further 110 jobs from Fawdon, Newcastle; recognises the losses come within a month of triggering Article 50, and as negotiations with the EU on the UK leaving the EU and the UK's future with the EU are commencing; further recognises the cost of importing products, including sugar, cocoa and production machinery, has risen due to the weakness of the pound and the uncertainty over the UK's future relationship with the single market and customs union; 

In [12]:
q= "Bob Smith was in the Houses of Parliament the other day"
entities(q)

Bob Smith was in the Houses of Parliament the other day
-------------- entities only ---------------
(Bob Smith, the Houses of Parliament, the other day)
{'Bob Smith': [(378, 'PERSON')], 'the Houses of Parliament': [(381, 'ORG')], 'the other day': [(388, 'DATE')]}


<H2>Stanford NER</H2>

In [15]:
import os
java_path = "C:/Program Files/Java/jdk1.8.0_151/bin"
os.environ['JAVA_HOME'] = java_path
#import nltk
#nltk.internals.config_java("C:/Program Files/Java/jdk1.6.0_30/bin/java.exe")

In [16]:
from nltk.tag import StanfordNERTagger
from nltk.tokenize import word_tokenize

st = StanfordNERTagger('E:/stanford-ner-2018-02-27/classifiers/english.all.3class.distsim.crf.ser.gz',
  'E:/stanford-ner-2018-02-27/stanford-ner.jar',encoding='utf-8')

text = 'While in France, Christine Lagarde discussed short-term stimulus efforts in a recent interview with the Wall Street Journal.'
k = "Tree is near the house"

tokenized_text = word_tokenize(k)
classified_text = st.tag(tokenized_text)

print(classified_text)

C:\Users\ADMIN\AppData\Local\Programs\Python\Python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)


[('Tree', 'O'), ('is', 'O'), ('near', 'O'), ('the', 'O'), ('house', 'O')]


<H2>Let's Code</H2>

I am using Standford NER for named entity recognition

<H2>Get all category names in quick draw dataset</H2>

In [17]:
category_file = open("categories.txt", "r")


In [18]:
category_lis = category_file.read().split('\n')

In [19]:
category_lis

['aircraft carrier',
 'airplane',
 'alarm clock',
 'ambulance',
 'angel',
 'animal migration',
 'ant',
 'anvil',
 'apple',
 'arm',
 'asparagus',
 'axe',
 'backpack',
 'banana',
 'bandage',
 'barn',
 'baseball',
 'baseball bat',
 'basket',
 'basketball',
 'bat',
 'bathtub',
 'beach',
 'bear',
 'beard',
 'bed',
 'bee',
 'belt',
 'bench',
 'bicycle',
 'binoculars',
 'bird',
 'birthday cake',
 'blackberry',
 'blueberry',
 'book',
 'boomerang',
 'bottlecap',
 'bowtie',
 'bracelet',
 'brain',
 'bread',
 'bridge',
 'broccoli',
 'broom',
 'bucket',
 'bulldozer',
 'bus',
 'bush',
 'butterfly',
 'cactus',
 'cake',
 'calculator',
 'calendar',
 'camel',
 'camera',
 'camouflage',
 'campfire',
 'candle',
 'cannon',
 'canoe',
 'car',
 'carrot',
 'castle',
 'cat',
 'ceiling fan',
 'cello',
 'cell phone',
 'chair',
 'chandelier',
 'church',
 'circle',
 'clarinet',
 'clock',
 'cloud',
 'coffee cup',
 'compass',
 'computer',
 'cookie',
 'cooler',
 'couch',
 'cow',
 'crab',
 'crayon',
 'crocodile',
 'crow

In [20]:
text = "WoW!!!! I love cats and dogs."

In [21]:
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
import string
def preprocess_text(text):
    tokens = word_tokenize(text)
    # convert to lower case
    tokens = [w.lower() for w in tokens]
    # remove punctuation from each word

    table = str.maketrans('', '', string.punctuation)
    stripped = [w.translate(table) for w in tokens]
    # remove remaining tokens that are not alphabetic
    words = [word for word in stripped if word.isalpha()]
    # filter out stop words

    stop_words = set(stopwords.words('english'))
    words = [w for w in words if not w in stop_words]
    return words


In [22]:
lis=preprocess_text(text)

In [23]:
lis

['wow', 'love', 'cats', 'dogs']

In [24]:
from fuzzywuzzy import fuzz
from fuzzywuzzy import process
process.extractOne('near', category_lis)


('ear', 86)

In [25]:
import nltk
from nltk.tag import StanfordNERTagger
from nltk.tokenize import word_tokenize
import sys
import warnings

warnings.simplefilter('always', DeprecationWarning)

def extract_entities(text):
    
    st = StanfordNERTagger('E:/stanford-ner-2018-02-27/classifiers/english.all.3class.distsim.crf.ser.gz',
                           'E:/stanford-ner-2018-02-27/stanford-ner.jar',encoding='utf-8')
    #tokenized_text = word_tokenize(text)
    
    classified_text = st.tag(text)
    return classified_text

In [26]:
sample="cats and dogs, I love them"

In [27]:
sample=preprocess_text(sample)

In [28]:
warnings.simplefilter('always', DeprecationWarning)
extracted_lis=extract_entities(sample)

C:\Users\ADMIN\AppData\Local\Programs\Python\Python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)


In [29]:
extracted_lis

[('cats', 'O'), ('dogs', 'O'), ('love', 'O')]

<H2>Load google pretrained word2vec model to match sementic similarity of the category list and the text descriptions</H2>

In [30]:
import gensim

# Load Google's pre-trained Word2Vec model.
model = gensim.models.KeyedVectors.load_word2vec_format('./GoogleNews-vectors-negative300.bin', binary=True) 

C:\Users\ADMIN\AppData\Local\Programs\Python\Python36\lib\site-packages\gensim\utils.py:1197: UserWarning: detected Windows; aliasing chunkize to chunkize_serial
  warnings.warn("detected Windows; aliasing chunkize to chunkize_serial")


In [20]:
model.similarity('trees','tree')

0.8293122426257508

In [34]:
import difflib
import sys
import warnings

if not sys.warnoptions:
    warnings.simplefilter("ignore")
def match_entities_with_dataset(text,category_lis):
    word_lis=preprocess_text(text)
    word_lis=extract_entities(word_lis)
    temp=[]
    for k in word_lis:
      t = difflib.get_close_matches(k[0], category_lis,1)
      if (t):
            if((t[0] in model.vocab) and (k[0] in model.vocab)and(model.similarity(k[0],t[0])>0.80)):
                
                temp.append(t[0])
    return temp

In [111]:
match_entities_with_dataset('trees near my House',category_lis)

c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)


['tree', 'house']

<H2> Let us use mcTest dataset to get the narrative text description</H2>

In [23]:
import pandas as pd
#df=pd.read_csv('mc500_train.csv', sep=',',header=None)

In [24]:
df=pd.read_csv('mc500_train.csv', sep=',',header=None)

In [25]:
df.head()

0                                                  1   \
0  mc500.train.0  Author: 2778399758;Work Time(s): 839;Qual. sco...   
1  mc500.train.1  Author: 3208685851;Work Time(s): 836;Qual. sco...   
2  mc500.train.2  Author: 3165469218;Work Time(s): 1546;Qual. sc...   
3  mc500.train.3  Author: 3323630398;Work Time(s): 712;Qual. sco...   
4  mc500.train.4  Author: 646501657;Work Time(s): 743;Qual. scor...   

                                                  2   \
0  Alyssa got to the beach after a long trip. She...   
1  One morning, Elena woke up, much like she did ...   
2  Billy was like a king on the school yard. A ki...   
3  Billy went to the farm to buy some beef for hi...   
4  The road to Grandpa's house was long and windi...   

                                                  3   \
0                       one: What city is Alyssa in?   
1  multiple: What is the very first thing Elena d...   
2                      multiple: Billy was like a...   
3               one: What did Billy buy at the farm?   
4         multiple: Why did Grandpa answer the door?   

                           4                                   5   \
0                        trip                               Miami   
1  she says hello to the tree  she throws the covers on the floor   
2                      prince                               dummy   
3                        Beef                             Chicken   
4  Because he saw the insects           Because Jimmy was walking   

                          6                                  7   \
0                    Atlanta                              beach   
1  she says hello to the sun                she gets out of bed   
2                      queen                               king   
3                       Cows                              Fence   
4      Because Jimmy knocked  Because the trip took a long time   

                                                  8                9   \
0                   one: Why did Alyssa go to Miami?             swim   
1  one: At what time of day does this story take ...  Before the tree   
2                           multiple: Who was Billy?   The skinny kid   
3   multiple: What color were the spots on the cows?             Blue   
4           one: Where do Jimmy and his Grandpa sit?       On insects   

             ...              \
0            ...               
1            ...               
2            ...               
3            ...               
4            ...               

                                                  13  \
0       multiple: How many friends does Alyssa have?   
1    multiple: What happened to Mr. Fish in the end?   
2          one: What did the boy grab from his home?   
3  multiple: How many chickens were on the purple...   
4  multiple: What was Jimmy's favorite insect in ...   

                                             14  \
0                                             1   
1  He got put back in the bowl, but he was dead   
2                                        a rock   
3                                           Ten   
4                                        Beetle   

                                           15                        16  \
0                                           2                         3   
1  He started swimming around in Elena's hand  He got put in a new bowl   
2                                     a bench                    a tree   
3                                         Six                      Five   
4                                    Lady bug                      Moth   

                                                  17  \
0                                                  4   
1  He got put back in his bowl and started swimmi...   
2                                       Fishing line   
3                                               Four   
4                                      Lightning bug   

                               

In [114]:
import warnings

with warnings.catch_warnings():
    warnings.filterwarnings("ignore",category=DeprecationWarning)
    from nltk.tag import StanfordNERTagger 
result=[]
for txt in df[2]:
    result.extend(match_entities_with_dataset(txt,category_lis))

c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\a

c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\a

c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\a

c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\a

c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\a

c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\a

c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\a

c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\a

c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\a

c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\a

c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\a

c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\a

c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\appdata\local\programs\python\python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)
c:\users\admin\a

In [115]:
##Final result after namming the named entities to category
result

['beach',
 'beach',
 'house',
 'house',
 'steak',
 'house',
 'bed',
 'tree',
 'tree',
 'fish',
 'fish',
 'fish',
 'castle',
 'fish',
 'fish',
 'face',
 'bench',
 'tree',
 'line',
 'line',
 'tree',
 'bench',
 'bench',
 'face',
 'line',
 'face',
 'saw',
 'saw',
 'fence',
 'fence',
 'house',
 'tree',
 'house',
 'house',
 'lightning',
 'house',
 'door',
 'lightning',
 'house',
 'house',
 'tree',
 'television',
 'television',
 'telephone',
 'lightning',
 'rain',
 'potato',
 'bed',
 'strawberry',
 'spider',
 'saw',
 'spider',
 'spider',
 'spider',
 'spider',
 'spider',
 'saw',
 'horse',
 'cow',
 'horse',
 'cow',
 'dog',
 'sandwich',
 'fish',
 'sandwich',
 'horse',
 'cow',
 'onion',
 'potato',
 'horse',
 'cow',
 'dog',
 'sandwich',
 'fish',
 'sandwich',
 'house',
 'face',
 'saw',
 'saw',
 'dog',
 'dog',
 'bicycle',
 'bicycle',
 'river',
 'house',
 'bicycle',
 'river',
 'river',
 'river',
 'tree',
 'house',
 'ladder',
 'tree',
 'house',
 'ladder',
 'bucket',
 'bucket',
 'ladder',
 'ladder',
 '

In [117]:
##Eliminating duplicates to check the number of categories identified
result_final = list(set(result))

In [118]:
len(result_final)

163

163 categories are detected. We can train the NER and word2vec model to improve this detection rate. For instance the categories like "school bus" and others that have two consecutive words are not detected by our model. This can be improved by training.

<H2>Demo</H2>

In [32]:
demo_txt = "The dog is near the Tree"

In [35]:
result=[]
result.extend(match_entities_with_dataset(demo_txt,category_lis))

C:\Users\ADMIN\AppData\Local\Programs\Python\Python36\lib\site-packages\nltk\tag\stanford.py:183: DeprecationWarning: 
The StanfordTokenizer will be deprecated in version 3.2.5.
Please use nltk.tag.corenlp.CoreNLPPOSTagger or nltk.tag.corenlp.CoreNLPNERTagger instead.
  super(StanfordNERTagger, self).__init__(*args, **kwargs)


In [36]:
result_final = list(set(result))

In [37]:
result_final

['dog', 'tree']

Let's consider a sample text from which the standford NER identifies the entities. We can then use the entities and do semantic matching with Google's pretrained word2vec model. Once we get the categories we can use the appropriate npy files as input to auto-encoder , do clustering, decode only the centroids and add it to the dictionary to be sent to the generate response module.

<H2>Let's use auto-encoder to convert each npy image to latent space and then decode to get the original values.</H2>

In [40]:
%matplotlib inline

import umap
import numpy as np
import pandas as pd
from scipy.stats import norm
from keras.layers import *
from keras.models import Model
import matplotlib.pyplot as plt

Using TensorFlow backend.


In [45]:
del data

In [43]:
data = np.load('data_npy/full%2Fnumpy_bitmap%2F'+result_final[0]+'.npy')

In [44]:
data.shape

(152159, 784)

In [46]:
encoding_size = 32

input_image = Input(shape=(28, 28, 1))
x = Conv2D(32, (3, 3), activation='relu', padding='same', data_format='channels_last')(input_image)
x = MaxPooling2D((2, 2), padding='same', data_format='channels_last')(x)
x = Dropout(0.2)(x)
x = Conv2D(32, (3, 3), activation='relu', padding='same', data_format='channels_last')(x)
x = MaxPooling2D((2, 2), padding='same', data_format='channels_last')(x)
x = Dropout(0.2)(x)
x = Conv2D(32, (3, 3), activation='relu', padding='same', data_format='channels_last')(x)
x = MaxPooling2D((2, 2), padding='same', data_format='channels_last')(x)

x = Flatten()(x)                                        # Squashing
x = Dense(256, activation='relu')(x)                    # Squashing
x = Dropout(0.2)(x)                                     # Squashing
encoded = Dense(encoding_size, activation='relu')(x)    # Dense, encoded layer
x = Dense(64*7*7, activation='relu')(encoded)           # Growing
x = Dropout(0.2)(x)                                     # Growing
x = Reshape((-1, 7, 7, 64)[1:])(x)                      # Growing

x = Conv2D(32, (3, 3), activation='relu', padding='same', data_format='channels_last')(x)
x = UpSampling2D((2, 2), data_format='channels_last')(x)
x = Conv2D(32, (3, 3), activation='relu', padding='same', data_format='channels_last')(x)
x = UpSampling2D((2, 2), data_format='channels_last')(x)
x = Conv2D(32, (3, 3), activation='relu', padding='same', data_format='channels_last')(x)
decoded = Conv2D(1, (3, 3), activation='sigmoid', padding='same', data_format='channels_last')(x)

autoencoder = Model(input_image, decoded)

In [47]:
autoencoder.compile(optimizer='adam', 
                    loss='binary_crossentropy', 
                    metrics=['mse'])

autoencoder.summary()


_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         (None, 28, 28, 1)         0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 28, 28, 32)        320       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 14, 14, 32)        0         
_________________________________________________________________
dropout_1 (Dropout)          (None, 14, 14, 32)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 14, 14, 32)        9248      
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 7, 7, 32)          0         
_________________________________________________________________
dropout_2 (Dropout)          (None, 7, 7, 32)          0         
__________

Let K value be 5 for clustering, thus returning 5 centroids

In [50]:
from sklearn.cluster import KMeans

In [51]:
lib_dict={}

In [52]:
for entity in result_final:
    data=np.load('data_npy/full%2Fnumpy_bitmap%2F'+entity+'.npy')
    data = data.reshape(-1, 28, 28, 1) / 255.0
    X = data[:5000]
    autoencoder.fit(X, X, 
                batch_size=32,
                validation_split=0.2,
                epochs=10)
    input_image = Input(shape=(28, 28, 1))
    x = Conv2D(32, (3, 3), activation='relu', padding='same', data_format='channels_last',
           weights=autoencoder.layers[1].get_weights())(input_image)
    x = MaxPooling2D((2, 2), padding='same', data_format='channels_last')(x)
    x = Dropout(0.2)(x)
    x = Conv2D(32, (3, 3), activation='relu', padding='same', data_format='channels_last',
           weights=autoencoder.layers[4].get_weights())(x)
    x = MaxPooling2D((2, 2), padding='same', data_format='channels_last')(x)
    x = Dropout(0.2)(x)
    x = Conv2D(32, (3, 3), activation='relu', padding='same', data_format='channels_last',
           weights=autoencoder.layers[7].get_weights())(x)
    x = MaxPooling2D((2, 2), padding='same', data_format='channels_last')(x)
    x = Flatten()(x)
    x = Dense(256, activation='relu',
          weights=autoencoder.layers[10].get_weights())(x)
    x = Dropout(0.2)(x)
    encoded = Dense(encoding_size, activation='relu',
          weights=autoencoder.layers[12].get_weights())(x)

    encoder = Model(input_image, encoded)
    latent_images = data.reshape(-1, 28, 28, 1)
    Z= encoder.predict(latent_images)
    ##Apply Kmeans
    # Number of clusters
    kmeans = KMeans(n_clusters=5)
    # Fitting the input data
    kmeans = kmeans.fit(Z)
    # Getting the cluster labels
    labels = kmeans.predict(Z)
    # Centroid values
    centroids = kmeans.cluster_centers_
    ##Decode
    latent_vector = Input(shape=(32,))
    y = Dense(64*7*7, activation='relu', 
          weights=autoencoder.layers[13].get_weights())(latent_vector)
    y = Reshape((-1, 7, 7, 64)[1:])(y)
    y = Conv2D(32, (3, 3), activation='relu', padding='same', data_format='channels_last', 
           weights=autoencoder.layers[16].get_weights())(y)
    y = UpSampling2D((2, 2), data_format='channels_last')(y)
    y = Conv2D(32, (3, 3), activation='relu', padding='same', data_format='channels_last', 
          weights=autoencoder.layers[18].get_weights())(y)
    y = UpSampling2D((2, 2), data_format='channels_last')(y)
    y = Conv2D(32, (3, 3), activation='relu', padding='same', data_format='channels_last', 
           weights=autoencoder.layers[20].get_weights())(y)
    decoded = Conv2D(1, (3, 3), activation='sigmoid', padding='same', data_format='channels_last', 
                 weights=autoencoder.layers[21].get_weights())(y)

    decoder = Model(latent_vector, decoded)
    strokes_decoded=decoder.predict(centroids)
    lib_dict[entity] = strokes_decoded

Train on 4000 samples, validate on 1000 samples
Epoch 1/10
4000/4000 [==============================] - ETA: 6:12:36 - loss: 0.6941 - mean_squared_error: 0.215 - ETA: 3:05:36 - loss: 0.6929 - mean_squared_error: 0.213 - ETA: 2:03:10 - loss: 0.6919 - mean_squared_error: 0.212 - ETA: 1:31:54 - loss: 0.6904 - mean_squared_error: 0.212 - ETA: 1:13:11 - loss: 0.6884 - mean_squared_error: 0.211 - ETA: 1:00:43 - loss: 0.6854 - mean_squared_error: 0.210 - ETA: 51:48 - loss: 0.6807 - mean_squared_error: 0.2085  - ETA: 45:05 - loss: 0.6741 - mean_squared_error: 0.205 - ETA: 39:50 - loss: 0.6642 - mean_squared_error: 0.200 - ETA: 35:38 - loss: 0.6521 - mean_squared_error: 0.194 - ETA: 32:12 - loss: 0.6441 - mean_squared_error: 0.189 - ETA: 29:19 - loss: 0.6363 - mean_squared_error: 0.184 - ETA: 26:53 - loss: 0.6301 - mean_squared_error: 0.181 - ETA: 24:48 - loss: 0.6225 - mean_squared_error: 0.177 - ETA: 22:59 - loss: 0.6147 - mean_squared_error: 0.174 - ETA: 21:24 - loss: 0.6090 - mean_squared_e

4000/4000 [==============================] - ETA: 45s - loss: 0.3281 - mean_squared_error: 0.077 - ETA: 44s - loss: 0.3443 - mean_squared_error: 0.081 - ETA: 44s - loss: 0.3456 - mean_squared_error: 0.081 - ETA: 44s - loss: 0.3372 - mean_squared_error: 0.079 - ETA: 43s - loss: 0.3356 - mean_squared_error: 0.078 - ETA: 43s - loss: 0.3344 - mean_squared_error: 0.078 - ETA: 43s - loss: 0.3339 - mean_squared_error: 0.078 - ETA: 43s - loss: 0.3340 - mean_squared_error: 0.078 - ETA: 43s - loss: 0.3337 - mean_squared_error: 0.078 - ETA: 42s - loss: 0.3358 - mean_squared_error: 0.078 - ETA: 42s - loss: 0.3357 - mean_squared_error: 0.078 - ETA: 42s - loss: 0.3350 - mean_squared_error: 0.078 - ETA: 41s - loss: 0.3325 - mean_squared_error: 0.077 - ETA: 41s - loss: 0.3344 - mean_squared_error: 0.078 - ETA: 40s - loss: 0.3354 - mean_squared_error: 0.078 - ETA: 40s - loss: 0.3358 - mean_squared_error: 0.078 - ETA: 40s - loss: 0.3372 - mean_squared_error: 0.079 - ETA: 39s - loss: 0.3366 - mean_square

4000/4000 [==============================] - ETA: 42s - loss: 0.3092 - mean_squared_error: 0.071 - ETA: 42s - loss: 0.3065 - mean_squared_error: 0.070 - ETA: 42s - loss: 0.3119 - mean_squared_error: 0.072 - ETA: 41s - loss: 0.3123 - mean_squared_error: 0.072 - ETA: 41s - loss: 0.3151 - mean_squared_error: 0.072 - ETA: 41s - loss: 0.3106 - mean_squared_error: 0.071 - ETA: 40s - loss: 0.3150 - mean_squared_error: 0.072 - ETA: 40s - loss: 0.3115 - mean_squared_error: 0.071 - ETA: 40s - loss: 0.3114 - mean_squared_error: 0.071 - ETA: 40s - loss: 0.3111 - mean_squared_error: 0.071 - ETA: 39s - loss: 0.3099 - mean_squared_error: 0.071 - ETA: 39s - loss: 0.3092 - mean_squared_error: 0.071 - ETA: 38s - loss: 0.3096 - mean_squared_error: 0.071 - ETA: 38s - loss: 0.3090 - mean_squared_error: 0.071 - ETA: 38s - loss: 0.3091 - mean_squared_error: 0.071 - ETA: 38s - loss: 0.3100 - mean_squared_error: 0.071 - ETA: 37s - loss: 0.3093 - mean_squared_error: 0.071 - ETA: 37s - loss: 0.3098 - mean_square

4000/4000 [==============================] - ETA: 44s - loss: 0.2886 - mean_squared_error: 0.065 - ETA: 44s - loss: 0.2999 - mean_squared_error: 0.067 - ETA: 43s - loss: 0.2965 - mean_squared_error: 0.066 - ETA: 43s - loss: 0.2917 - mean_squared_error: 0.066 - ETA: 43s - loss: 0.2951 - mean_squared_error: 0.066 - ETA: 43s - loss: 0.2944 - mean_squared_error: 0.066 - ETA: 42s - loss: 0.2945 - mean_squared_error: 0.066 - ETA: 42s - loss: 0.2948 - mean_squared_error: 0.066 - ETA: 41s - loss: 0.2957 - mean_squared_error: 0.066 - ETA: 41s - loss: 0.2953 - mean_squared_error: 0.066 - ETA: 41s - loss: 0.2966 - mean_squared_error: 0.067 - ETA: 40s - loss: 0.2973 - mean_squared_error: 0.067 - ETA: 40s - loss: 0.2986 - mean_squared_error: 0.067 - ETA: 40s - loss: 0.2991 - mean_squared_error: 0.067 - ETA: 39s - loss: 0.3003 - mean_squared_error: 0.067 - ETA: 39s - loss: 0.3005 - mean_squared_error: 0.067 - ETA: 39s - loss: 0.3005 - mean_squared_error: 0.067 - ETA: 38s - loss: 0.3007 - mean_square

4000/4000 [==============================] - ETA: 46s - loss: 0.2608 - mean_squared_error: 0.057 - ETA: 46s - loss: 0.2694 - mean_squared_error: 0.059 - ETA: 45s - loss: 0.2694 - mean_squared_error: 0.059 - ETA: 44s - loss: 0.2731 - mean_squared_error: 0.059 - ETA: 43s - loss: 0.2733 - mean_squared_error: 0.060 - ETA: 43s - loss: 0.2796 - mean_squared_error: 0.061 - ETA: 42s - loss: 0.2838 - mean_squared_error: 0.062 - ETA: 42s - loss: 0.2861 - mean_squared_error: 0.063 - ETA: 41s - loss: 0.2848 - mean_squared_error: 0.062 - ETA: 41s - loss: 0.2852 - mean_squared_error: 0.063 - ETA: 40s - loss: 0.2856 - mean_squared_error: 0.063 - ETA: 40s - loss: 0.2878 - mean_squared_error: 0.063 - ETA: 40s - loss: 0.2876 - mean_squared_error: 0.063 - ETA: 40s - loss: 0.2877 - mean_squared_error: 0.063 - ETA: 40s - loss: 0.2866 - mean_squared_error: 0.063 - ETA: 39s - loss: 0.2870 - mean_squared_error: 0.063 - ETA: 39s - loss: 0.2880 - mean_squared_error: 0.063 - ETA: 38s - loss: 0.2889 - mean_square

4000/4000 [==============================] - ETA: 48s - loss: 0.2983 - mean_squared_error: 0.065 - ETA: 45s - loss: 0.2978 - mean_squared_error: 0.064 - ETA: 46s - loss: 0.2962 - mean_squared_error: 0.064 - ETA: 44s - loss: 0.2959 - mean_squared_error: 0.065 - ETA: 42s - loss: 0.2931 - mean_squared_error: 0.064 - ETA: 41s - loss: 0.2925 - mean_squared_error: 0.064 - ETA: 41s - loss: 0.2929 - mean_squared_error: 0.064 - ETA: 40s - loss: 0.2921 - mean_squared_error: 0.064 - ETA: 40s - loss: 0.2912 - mean_squared_error: 0.063 - ETA: 39s - loss: 0.2899 - mean_squared_error: 0.063 - ETA: 39s - loss: 0.2910 - mean_squared_error: 0.063 - ETA: 39s - loss: 0.2901 - mean_squared_error: 0.063 - ETA: 39s - loss: 0.2893 - mean_squared_error: 0.063 - ETA: 39s - loss: 0.2888 - mean_squared_error: 0.063 - ETA: 38s - loss: 0.2880 - mean_squared_error: 0.063 - ETA: 38s - loss: 0.2878 - mean_squared_error: 0.063 - ETA: 37s - loss: 0.2875 - mean_squared_error: 0.063 - ETA: 37s - loss: 0.2871 - mean_square

4000/4000 [==============================] - ETA: 43s - loss: 0.2804 - mean_squared_error: 0.059 - ETA: 43s - loss: 0.2712 - mean_squared_error: 0.057 - ETA: 44s - loss: 0.2734 - mean_squared_error: 0.058 - ETA: 44s - loss: 0.2777 - mean_squared_error: 0.059 - ETA: 44s - loss: 0.2787 - mean_squared_error: 0.059 - ETA: 44s - loss: 0.2796 - mean_squared_error: 0.060 - ETA: 43s - loss: 0.2810 - mean_squared_error: 0.060 - ETA: 43s - loss: 0.2797 - mean_squared_error: 0.060 - ETA: 43s - loss: 0.2800 - mean_squared_error: 0.060 - ETA: 42s - loss: 0.2794 - mean_squared_error: 0.060 - ETA: 42s - loss: 0.2786 - mean_squared_error: 0.060 - ETA: 41s - loss: 0.2795 - mean_squared_error: 0.060 - ETA: 41s - loss: 0.2797 - mean_squared_error: 0.060 - ETA: 40s - loss: 0.2803 - mean_squared_error: 0.060 - ETA: 40s - loss: 0.2791 - mean_squared_error: 0.060 - ETA: 39s - loss: 0.2779 - mean_squared_error: 0.059 - ETA: 39s - loss: 0.2787 - mean_squared_error: 0.059 - ETA: 39s - loss: 0.2774 - mean_square

4000/4000 [==============================] - ETA: 38s - loss: 0.2839 - mean_squared_error: 0.060 - ETA: 40s - loss: 0.2733 - mean_squared_error: 0.057 - ETA: 39s - loss: 0.2808 - mean_squared_error: 0.059 - ETA: 39s - loss: 0.2785 - mean_squared_error: 0.059 - ETA: 39s - loss: 0.2731 - mean_squared_error: 0.058 - ETA: 39s - loss: 0.2759 - mean_squared_error: 0.059 - ETA: 38s - loss: 0.2773 - mean_squared_error: 0.059 - ETA: 38s - loss: 0.2790 - mean_squared_error: 0.060 - ETA: 38s - loss: 0.2781 - mean_squared_error: 0.059 - ETA: 38s - loss: 0.2785 - mean_squared_error: 0.059 - ETA: 37s - loss: 0.2793 - mean_squared_error: 0.060 - ETA: 37s - loss: 0.2782 - mean_squared_error: 0.059 - ETA: 37s - loss: 0.2777 - mean_squared_error: 0.059 - ETA: 37s - loss: 0.2788 - mean_squared_error: 0.059 - ETA: 36s - loss: 0.2786 - mean_squared_error: 0.059 - ETA: 36s - loss: 0.2792 - mean_squared_error: 0.060 - ETA: 36s - loss: 0.2789 - mean_squared_error: 0.059 - ETA: 36s - loss: 0.2785 - mean_square

4000/4000 [==============================] - ETA: 47s - loss: 0.2821 - mean_squared_error: 0.061 - ETA: 47s - loss: 0.2810 - mean_squared_error: 0.060 - ETA: 47s - loss: 0.2710 - mean_squared_error: 0.057 - ETA: 46s - loss: 0.2690 - mean_squared_error: 0.057 - ETA: 45s - loss: 0.2684 - mean_squared_error: 0.056 - ETA: 44s - loss: 0.2692 - mean_squared_error: 0.057 - ETA: 45s - loss: 0.2694 - mean_squared_error: 0.057 - ETA: 44s - loss: 0.2692 - mean_squared_error: 0.057 - ETA: 44s - loss: 0.2679 - mean_squared_error: 0.056 - ETA: 44s - loss: 0.2674 - mean_squared_error: 0.056 - ETA: 43s - loss: 0.2686 - mean_squared_error: 0.056 - ETA: 43s - loss: 0.2703 - mean_squared_error: 0.057 - ETA: 43s - loss: 0.2696 - mean_squared_error: 0.057 - ETA: 42s - loss: 0.2695 - mean_squared_error: 0.057 - ETA: 41s - loss: 0.2704 - mean_squared_error: 0.057 - ETA: 41s - loss: 0.2698 - mean_squared_error: 0.057 - ETA: 40s - loss: 0.2695 - mean_squared_error: 0.057 - ETA: 39s - loss: 0.2687 - mean_square

4000/4000 [==============================] - ETA: 45s - loss: 0.2700 - mean_squared_error: 0.057 - ETA: 45s - loss: 0.2604 - mean_squared_error: 0.054 - ETA: 46s - loss: 0.2600 - mean_squared_error: 0.054 - ETA: 46s - loss: 0.2650 - mean_squared_error: 0.055 - ETA: 46s - loss: 0.2654 - mean_squared_error: 0.056 - ETA: 46s - loss: 0.2638 - mean_squared_error: 0.055 - ETA: 45s - loss: 0.2645 - mean_squared_error: 0.055 - ETA: 44s - loss: 0.2653 - mean_squared_error: 0.055 - ETA: 44s - loss: 0.2650 - mean_squared_error: 0.055 - ETA: 43s - loss: 0.2646 - mean_squared_error: 0.055 - ETA: 43s - loss: 0.2648 - mean_squared_error: 0.055 - ETA: 42s - loss: 0.2673 - mean_squared_error: 0.056 - ETA: 42s - loss: 0.2672 - mean_squared_error: 0.056 - ETA: 42s - loss: 0.2680 - mean_squared_error: 0.056 - ETA: 41s - loss: 0.2688 - mean_squared_error: 0.056 - ETA: 41s - loss: 0.2690 - mean_squared_error: 0.056 - ETA: 41s - loss: 0.2707 - mean_squared_error: 0.057 - ETA: 40s - loss: 0.2711 - mean_square

4000/4000 [==============================] - ETA: 53s - loss: 0.3188 - mean_squared_error: 0.074 - ETA: 56s - loss: 0.3127 - mean_squared_error: 0.070 - ETA: 53s - loss: 0.3062 - mean_squared_error: 0.069 - ETA: 52s - loss: 0.3064 - mean_squared_error: 0.068 - ETA: 52s - loss: 0.3061 - mean_squared_error: 0.068 - ETA: 51s - loss: 0.3028 - mean_squared_error: 0.067 - ETA: 51s - loss: 0.2998 - mean_squared_error: 0.066 - ETA: 49s - loss: 0.2993 - mean_squared_error: 0.066 - ETA: 48s - loss: 0.2973 - mean_squared_error: 0.066 - ETA: 47s - loss: 0.2963 - mean_squared_error: 0.065 - ETA: 46s - loss: 0.2970 - mean_squared_error: 0.065 - ETA: 47s - loss: 0.2964 - mean_squared_error: 0.065 - ETA: 46s - loss: 0.2972 - mean_squared_error: 0.065 - ETA: 46s - loss: 0.2975 - mean_squared_error: 0.065 - ETA: 45s - loss: 0.2970 - mean_squared_error: 0.065 - ETA: 44s - loss: 0.2969 - mean_squared_error: 0.065 - ETA: 43s - loss: 0.2970 - mean_squared_error: 0.065 - ETA: 43s - loss: 0.2961 - mean_square

4000/4000 [==============================] - ETA: 44s - loss: 0.2742 - mean_squared_error: 0.057 - ETA: 44s - loss: 0.2675 - mean_squared_error: 0.054 - ETA: 44s - loss: 0.2651 - mean_squared_error: 0.054 - ETA: 43s - loss: 0.2627 - mean_squared_error: 0.053 - ETA: 43s - loss: 0.2624 - mean_squared_error: 0.053 - ETA: 42s - loss: 0.2595 - mean_squared_error: 0.053 - ETA: 42s - loss: 0.2626 - mean_squared_error: 0.053 - ETA: 42s - loss: 0.2607 - mean_squared_error: 0.053 - ETA: 42s - loss: 0.2603 - mean_squared_error: 0.053 - ETA: 41s - loss: 0.2599 - mean_squared_error: 0.052 - ETA: 41s - loss: 0.2589 - mean_squared_error: 0.052 - ETA: 40s - loss: 0.2584 - mean_squared_error: 0.052 - ETA: 40s - loss: 0.2572 - mean_squared_error: 0.052 - ETA: 40s - loss: 0.2567 - mean_squared_error: 0.052 - ETA: 40s - loss: 0.2562 - mean_squared_error: 0.051 - ETA: 39s - loss: 0.2563 - mean_squared_error: 0.051 - ETA: 39s - loss: 0.2570 - mean_squared_error: 0.052 - ETA: 38s - loss: 0.2573 - mean_square

4000/4000 [==============================] - ETA: 42s - loss: 0.2529 - mean_squared_error: 0.050 - ETA: 45s - loss: 0.2499 - mean_squared_error: 0.049 - ETA: 44s - loss: 0.2474 - mean_squared_error: 0.048 - ETA: 44s - loss: 0.2484 - mean_squared_error: 0.049 - ETA: 43s - loss: 0.2490 - mean_squared_error: 0.049 - ETA: 42s - loss: 0.2496 - mean_squared_error: 0.049 - ETA: 42s - loss: 0.2500 - mean_squared_error: 0.049 - ETA: 41s - loss: 0.2505 - mean_squared_error: 0.049 - ETA: 41s - loss: 0.2520 - mean_squared_error: 0.049 - ETA: 40s - loss: 0.2524 - mean_squared_error: 0.049 - ETA: 40s - loss: 0.2530 - mean_squared_error: 0.050 - ETA: 40s - loss: 0.2517 - mean_squared_error: 0.049 - ETA: 39s - loss: 0.2525 - mean_squared_error: 0.049 - ETA: 39s - loss: 0.2530 - mean_squared_error: 0.049 - ETA: 39s - loss: 0.2529 - mean_squared_error: 0.049 - ETA: 38s - loss: 0.2532 - mean_squared_error: 0.050 - ETA: 38s - loss: 0.2530 - mean_squared_error: 0.050 - ETA: 37s - loss: 0.2535 - mean_square

4000/4000 [==============================] - ETA: 44s - loss: 0.2498 - mean_squared_error: 0.049 - ETA: 46s - loss: 0.2554 - mean_squared_error: 0.051 - ETA: 47s - loss: 0.2484 - mean_squared_error: 0.048 - ETA: 47s - loss: 0.2461 - mean_squared_error: 0.048 - ETA: 47s - loss: 0.2435 - mean_squared_error: 0.047 - ETA: 46s - loss: 0.2461 - mean_squared_error: 0.048 - ETA: 45s - loss: 0.2451 - mean_squared_error: 0.047 - ETA: 45s - loss: 0.2456 - mean_squared_error: 0.047 - ETA: 45s - loss: 0.2457 - mean_squared_error: 0.047 - ETA: 44s - loss: 0.2459 - mean_squared_error: 0.048 - ETA: 44s - loss: 0.2458 - mean_squared_error: 0.048 - ETA: 43s - loss: 0.2457 - mean_squared_error: 0.048 - ETA: 43s - loss: 0.2451 - mean_squared_error: 0.047 - ETA: 43s - loss: 0.2449 - mean_squared_error: 0.047 - ETA: 42s - loss: 0.2452 - mean_squared_error: 0.047 - ETA: 42s - loss: 0.2455 - mean_squared_error: 0.047 - ETA: 41s - loss: 0.2447 - mean_squared_error: 0.047 - ETA: 41s - loss: 0.2444 - mean_square

4000/4000 [==============================] - ETA: 45s - loss: 0.2438 - mean_squared_error: 0.046 - ETA: 45s - loss: 0.2380 - mean_squared_error: 0.045 - ETA: 44s - loss: 0.2415 - mean_squared_error: 0.046 - ETA: 44s - loss: 0.2388 - mean_squared_error: 0.045 - ETA: 43s - loss: 0.2398 - mean_squared_error: 0.045 - ETA: 43s - loss: 0.2388 - mean_squared_error: 0.045 - ETA: 43s - loss: 0.2410 - mean_squared_error: 0.046 - ETA: 42s - loss: 0.2415 - mean_squared_error: 0.046 - ETA: 42s - loss: 0.2425 - mean_squared_error: 0.046 - ETA: 42s - loss: 0.2423 - mean_squared_error: 0.046 - ETA: 41s - loss: 0.2416 - mean_squared_error: 0.046 - ETA: 40s - loss: 0.2420 - mean_squared_error: 0.046 - ETA: 40s - loss: 0.2431 - mean_squared_error: 0.046 - ETA: 40s - loss: 0.2423 - mean_squared_error: 0.046 - ETA: 39s - loss: 0.2426 - mean_squared_error: 0.046 - ETA: 39s - loss: 0.2429 - mean_squared_error: 0.046 - ETA: 39s - loss: 0.2431 - mean_squared_error: 0.046 - ETA: 38s - loss: 0.2437 - mean_square

4000/4000 [==============================] - ETA: 44s - loss: 0.2298 - mean_squared_error: 0.043 - ETA: 45s - loss: 0.2349 - mean_squared_error: 0.044 - ETA: 46s - loss: 0.2346 - mean_squared_error: 0.044 - ETA: 46s - loss: 0.2378 - mean_squared_error: 0.044 - ETA: 46s - loss: 0.2360 - mean_squared_error: 0.044 - ETA: 45s - loss: 0.2363 - mean_squared_error: 0.044 - ETA: 45s - loss: 0.2378 - mean_squared_error: 0.045 - ETA: 45s - loss: 0.2387 - mean_squared_error: 0.045 - ETA: 44s - loss: 0.2385 - mean_squared_error: 0.045 - ETA: 44s - loss: 0.2393 - mean_squared_error: 0.045 - ETA: 43s - loss: 0.2393 - mean_squared_error: 0.045 - ETA: 43s - loss: 0.2404 - mean_squared_error: 0.045 - ETA: 43s - loss: 0.2404 - mean_squared_error: 0.045 - ETA: 42s - loss: 0.2401 - mean_squared_error: 0.045 - ETA: 42s - loss: 0.2401 - mean_squared_error: 0.045 - ETA: 42s - loss: 0.2411 - mean_squared_error: 0.046 - ETA: 41s - loss: 0.2408 - mean_squared_error: 0.045 - ETA: 41s - loss: 0.2407 - mean_square

4000/4000 [==============================] - ETA: 45s - loss: 0.2310 - mean_squared_error: 0.043 - ETA: 43s - loss: 0.2282 - mean_squared_error: 0.042 - ETA: 43s - loss: 0.2304 - mean_squared_error: 0.042 - ETA: 43s - loss: 0.2341 - mean_squared_error: 0.043 - ETA: 43s - loss: 0.2389 - mean_squared_error: 0.045 - ETA: 42s - loss: 0.2398 - mean_squared_error: 0.045 - ETA: 42s - loss: 0.2391 - mean_squared_error: 0.045 - ETA: 41s - loss: 0.2402 - mean_squared_error: 0.045 - ETA: 41s - loss: 0.2408 - mean_squared_error: 0.045 - ETA: 40s - loss: 0.2407 - mean_squared_error: 0.045 - ETA: 40s - loss: 0.2409 - mean_squared_error: 0.045 - ETA: 40s - loss: 0.2411 - mean_squared_error: 0.045 - ETA: 39s - loss: 0.2408 - mean_squared_error: 0.045 - ETA: 39s - loss: 0.2418 - mean_squared_error: 0.045 - ETA: 39s - loss: 0.2416 - mean_squared_error: 0.045 - ETA: 38s - loss: 0.2423 - mean_squared_error: 0.046 - ETA: 38s - loss: 0.2413 - mean_squared_error: 0.045 - ETA: 37s - loss: 0.2408 - mean_square

4000/4000 [==============================] - ETA: 44s - loss: 0.2408 - mean_squared_error: 0.044 - ETA: 44s - loss: 0.2397 - mean_squared_error: 0.044 - ETA: 43s - loss: 0.2346 - mean_squared_error: 0.043 - ETA: 42s - loss: 0.2378 - mean_squared_error: 0.044 - ETA: 42s - loss: 0.2436 - mean_squared_error: 0.045 - ETA: 42s - loss: 0.2437 - mean_squared_error: 0.045 - ETA: 41s - loss: 0.2425 - mean_squared_error: 0.045 - ETA: 41s - loss: 0.2440 - mean_squared_error: 0.046 - ETA: 41s - loss: 0.2417 - mean_squared_error: 0.045 - ETA: 41s - loss: 0.2407 - mean_squared_error: 0.045 - ETA: 41s - loss: 0.2406 - mean_squared_error: 0.045 - ETA: 40s - loss: 0.2414 - mean_squared_error: 0.045 - ETA: 40s - loss: 0.2426 - mean_squared_error: 0.045 - ETA: 39s - loss: 0.2421 - mean_squared_error: 0.045 - ETA: 39s - loss: 0.2416 - mean_squared_error: 0.045 - ETA: 39s - loss: 0.2413 - mean_squared_error: 0.045 - ETA: 38s - loss: 0.2409 - mean_squared_error: 0.045 - ETA: 38s - loss: 0.2407 - mean_square

4000/4000 [==============================] - ETA: 46s - loss: 0.2419 - mean_squared_error: 0.045 - ETA: 45s - loss: 0.2382 - mean_squared_error: 0.044 - ETA: 45s - loss: 0.2358 - mean_squared_error: 0.044 - ETA: 44s - loss: 0.2317 - mean_squared_error: 0.042 - ETA: 44s - loss: 0.2313 - mean_squared_error: 0.042 - ETA: 43s - loss: 0.2307 - mean_squared_error: 0.042 - ETA: 43s - loss: 0.2281 - mean_squared_error: 0.041 - ETA: 42s - loss: 0.2292 - mean_squared_error: 0.042 - ETA: 42s - loss: 0.2303 - mean_squared_error: 0.042 - ETA: 41s - loss: 0.2299 - mean_squared_error: 0.042 - ETA: 41s - loss: 0.2313 - mean_squared_error: 0.042 - ETA: 40s - loss: 0.2323 - mean_squared_error: 0.043 - ETA: 40s - loss: 0.2323 - mean_squared_error: 0.043 - ETA: 40s - loss: 0.2315 - mean_squared_error: 0.042 - ETA: 39s - loss: 0.2314 - mean_squared_error: 0.042 - ETA: 39s - loss: 0.2310 - mean_squared_error: 0.042 - ETA: 39s - loss: 0.2313 - mean_squared_error: 0.042 - ETA: 38s - loss: 0.2313 - mean_square

4000/4000 [==============================] - ETA: 47s - loss: 0.2354 - mean_squared_error: 0.044 - ETA: 49s - loss: 0.2350 - mean_squared_error: 0.043 - ETA: 49s - loss: 0.2337 - mean_squared_error: 0.043 - ETA: 48s - loss: 0.2358 - mean_squared_error: 0.044 - ETA: 48s - loss: 0.2358 - mean_squared_error: 0.044 - ETA: 47s - loss: 0.2339 - mean_squared_error: 0.043 - ETA: 48s - loss: 0.2348 - mean_squared_error: 0.043 - ETA: 47s - loss: 0.2360 - mean_squared_error: 0.044 - ETA: 47s - loss: 0.2358 - mean_squared_error: 0.044 - ETA: 46s - loss: 0.2366 - mean_squared_error: 0.044 - ETA: 46s - loss: 0.2362 - mean_squared_error: 0.044 - ETA: 45s - loss: 0.2363 - mean_squared_error: 0.044 - ETA: 45s - loss: 0.2364 - mean_squared_error: 0.044 - ETA: 44s - loss: 0.2363 - mean_squared_error: 0.044 - ETA: 44s - loss: 0.2359 - mean_squared_error: 0.044 - ETA: 43s - loss: 0.2366 - mean_squared_error: 0.044 - ETA: 43s - loss: 0.2369 - mean_squared_error: 0.044 - ETA: 43s - loss: 0.2368 - mean_square

In [53]:
lib_dict

{'dog': array([[[[2.04363384e-11],
          [1.27295525e-14],
          [2.55106657e-16],
          ...,
          [6.03020292e-13],
          [1.75641876e-10],
          [4.01530578e-08]],
 
         [[5.02509770e-11],
          [3.46778487e-13],
          [9.05178032e-14],
          ...,
          [1.38650766e-11],
          [1.96561059e-10],
          [2.32744490e-10]],
 
         [[1.00888568e-11],
          [5.77577959e-12],
          [3.07689789e-12],
          ...,
          [1.76104117e-10],
          [1.75823633e-09],
          [2.06891587e-10]],
 
         ...,
 
         [[1.49486212e-09],
          [5.72577610e-11],
          [1.90213505e-11],
          ...,
          [8.74880238e-11],
          [2.93954666e-10],
          [4.00965199e-11]],
 
         [[2.58472461e-08],
          [1.42448178e-10],
          [1.49440738e-11],
          ...,
          [2.40458931e-10],
          [2.22951435e-09],
          [2.91783508e-09]],
 
         [[2.68023814e-06],
          [1.673872

Let's get the tree entity centroids and draw the images

In [54]:
K_centroids = lib_dict.get("tree")

In [57]:
K_centroids.shape

(5, 28, 28, 1)

In [68]:
import numpy as np

# Save
np.save('lib_dict.npy', lib_dict) 

In [69]:
# Load
read_dictionary = np.load('lib_dict.npy').item()
print(read_dictionary['tree']) # displays "world"


[[[[1.71729719e-09]
   [7.48089299e-11]
   [8.32307313e-12]
   ...
   [9.04680150e-12]
   [4.74702534e-11]
   [1.46332425e-07]]

  [[1.36726644e-10]
   [2.68368522e-10]
   [2.34751635e-10]
   ...
   [5.64527203e-10]
   [5.55639368e-10]
   [1.61560398e-09]]

  [[1.27150536e-12]
   [2.31827377e-10]
   [1.46914547e-09]
   ...
   [8.32708658e-09]
   [4.86112983e-10]
   [3.39323347e-10]]

  ...

  [[1.73876503e-11]
   [1.82397841e-09]
   [2.39571851e-09]
   ...
   [1.28762041e-08]
   [3.30272920e-09]
   [8.28861335e-09]]

  [[3.22754296e-10]
   [1.16402274e-08]
   [1.42796512e-08]
   ...
   [4.37312941e-08]
   [2.19610143e-08]
   [2.74244286e-07]]

  [[1.32516277e-07]
   [3.00441272e-08]
   [1.17419559e-08]
   ...
   [2.67999845e-09]
   [1.23216159e-09]
   [2.97691276e-08]]]


 [[[7.22950091e-08]
   [3.15602144e-09]
   [1.48678583e-10]
   ...
   [3.63357846e-12]
   [5.44957551e-11]
   [6.47072568e-07]]

  [[2.63780606e-08]
   [3.35552812e-08]
   [1.12381322e-08]
   ...
   [3.55289853e-10]
 